src/engine/PDKind.h

```diff
@@ -0,0 +1,220 @@
+struct IFrameElementCompare {
+        bool operator()(const IFrameElement &a, const IFrameElement &b) const {
+            if (a.lemma == b.lemma) {
+                return a.counter_example.ctx < b.counter_example.ctx;
+            } else {
+                return a.lemma < b.lemma;
+            }
+        }
+};
+
+/**
+ * Wrapper for Induction Frame set.
+ */
+struct InductionFrame {
+private:
+    std::set<IFrameElement, IFrameElementCompare> elements;
+
+public:
+    InductionFrame() = default;
+
+    InductionFrame(const InductionFrame& other) : elements(other.elements) {}
+
+    InductionFrame(InductionFrame&& other) noexcept : elements(std::move(other.elements)) {}
+
+    InductionFrame& operator=(const InductionFrame& other) {
+        if (this != &other) {
+            elements = other.elements;
+        }
+        return *this;
+    }
+
+    InductionFrame& operator=(InductionFrame&& other) noexcept {
+        if (this != &other) {
+            elements = std::move(other.elements);
+        }
+        return *this;
+    }
+
+    bool operator==(const InductionFrame& other) {
+        return elements == other.elements;
+    }
+
+    ~InductionFrame() = default;
+
+    using iterator = std::set<IFrameElement, IFrameElementCompare>::iterator;
+    using const_iterator = std::set<IFrameElement, IFrameElementCompare>::const_iterator;
+
+    iterator begin() { return elements.begin(); }
+    const_iterator begin() const { return elements.begin(); }
+    const_iterator cbegin() const { return elements.cbegin(); }
+
+    iterator end() { return elements.end(); }
+    const_iterator end() const { return elements.end(); }
+    const_iterator cend() const { return elements.cend(); }
+
+    std::pair<iterator, bool> insert(const IFrameElement& element) {
+        return elements.insert(element);
+    }
+
+    size_t erase(const IFrameElement& element) {
+        return elements.erase(element);
+    }
+
+    iterator find(const IFrameElement& element) {
+        return elements.find(element);
+    }
+    const_iterator find(const IFrameElement& element) const {
+        return elements.find(element);
+    }
+
+    void clear() {
+        elements.clear();
+    }
+
+    size_t size() const {
+        return elements.size();
+    }
+
+    bool empty() const {
+        return elements.empty();
+    }
+};
+
+/**
+ * Wrapper for return value of the Push function.
+ */
+struct PushResult {
+    InductionFrame i_frame;
+    InductionFrame new_i_frame;
+    int n;
+    bool is_invalid;
+    int steps_to_ctx;
+    PushResult(InductionFrame i_frame,
+               InductionFrame new_i_frame,
+               int n,
+               bool is_invalid,
+               int steps_to_ctx) : i_frame(std::move(i_frame)), new_i_frame(std::move(new_i_frame)), n(n), is_invalid(is_invalid), steps_to_ctx(steps_to_ctx) {}
+};
+
+/**
+ * A data structure where r[i] represents the states that are reachable in i steps from some initial states, i.e., r[0].
+ */
+class RFrame {
```

Comment:
```suggestion
class RFrames {
```

This represents all the reachability frames, no? Plural seems more appropriate here.